Task 30 Camera rework

danielolssonIT opened the pull request:
Task 30 implementation from Group 1 (Daniel and Erik), including the three subtasks:

- Create a replacement for the InterpolatedCamera class (as it is being deprecated)
- Move the CamCtl responsibilities into the camera class itself
- Change the way the locked on orbit camera position is calculated so that the camera can operate without adding extra nodes into the tree

The camera behavior is slightly changed from before.  Please run the code in order to see the current behavior and provide us with feedback.

RuthgerD commented:
Initial impressions:
- Camera smoothing is a bit over the top, we probably want it a bit stiffer
- Switching from free to orbit doesn't preserve the angle (It always wants to look straight down)
- Free cam is very wiggly

Code quality is better than what it replaces so I wont bikeshed that.

I'll look into the camera wiggling myself, I remember fixing it at some point :thinking:

RuthgerD commented on project/src/utilities/controllable_camera/ControllableCamera.gd:
```diff
@@ -15,75 +15,58 @@
-var rot_x = 0
-var rot_y = 0
-var lookaround_speed = 0.01
+const LockedCam = preload("LockedCam.gd")
+const FreeCam = preload("FreeCam.gd")
```
prefer giving these scripts a class_name so that you can directly refer to them

danielolssonIT commented:
> Initial impressions:
> 
> * Camera smoothing is a bit over the top, we probably want it a bit stiffer
> * Switching from free to orbit doesn't preserve the angle (It always wants to look straight down)
> * Free cam is very wiggly
> 
> Code quality is better than what it replaces so I wont bikeshed that.
> 
> I'll look into the camera wiggling myself, I remember fixing it at some point 🤔

I have updated the pull request with a few new commits containing some improvements. I added the "hack" to make the camera less wiggly and also found a neat but incomprehensible solution to preserving the angle. It seems to work from different target rotations as well, such as steep terrain or target being upside down.

RuthgerD commented:
Camera feels incredibly satisfying! awesome job;
especially the transition between orbit and free working like the "perfect" solution I described.

Will check the code for anything weird

RuthgerD reviewed (approved):
I have a few nitpicks here and there, but it's mostly just a case of gdscript being a bad language so I won't bore you with that.

Dimitris told me to not merge anything till the course is over (since more people are working on the same thing) but I think I will incorporate this in one way or another.

danielolssonIT commented:
Sounds great! Hopefully there are some implementation ideas in this PR that could be useful to you.
Thanks for all the support and feedback on this task! 🙂

AeroStun reviewed (approved):
Love the smooth transition

RuthgerD commented:
Thanks!
